iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/VoltageAngleLimitAdderImpl.java

```diff
@@ -0,0 +1,117 @@
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+package com.powsybl.iidm.network.impl;
+
+import java.util.Optional;
+
+import com.powsybl.iidm.network.*;
+import com.powsybl.iidm.network.TerminalRef.Side;
+import com.powsybl.iidm.network.VoltageAngleLimit.FlowDirection;
+import com.powsybl.iidm.network.impl.util.Ref;
+
+/**
+ *
+ * @author Luma Zamarreño <zamarrenolm at aia.es>
+ * @author José Antonio Marqués <marquesja at aia.es>
+ */
+class VoltageAngleLimitAdderImpl implements VoltageAngleLimitAdder {
+
+    private final Ref<NetworkImpl> networkRef;
+    private TerminalRef from;
+    private TerminalRef to;
+    private double limit = Double.NaN;
+    private FlowDirection flowDirection = FlowDirection.BOTH_DIRECTIONS;
+
+    VoltageAngleLimitAdderImpl(Ref<NetworkImpl> networkRef) {
+        this.networkRef = networkRef;
+    }
+
+    @Override
+    public VoltageAngleLimitAdderImpl from(TerminalRef from) {
+        this.from = from;
+        return this;
+    }
+
+    @Override
+    public VoltageAngleLimitAdderImpl to(TerminalRef to) {
+        this.to = to;
+        return this;
+    }
+
+    @Override
+    public VoltageAngleLimitAdderImpl withLimit(double limit) {
+        this.limit = limit;
+        return this;
+    }
+
+    @Override
+    public VoltageAngleLimitAdderImpl withFlowDirection(FlowDirection flowDirection) {
+        this.flowDirection = flowDirection;
+        return this;
+    }
+
+    @Override
+    public VoltageAngleLimit add() {
+        if (limit <= 0) {
+            throw new IllegalStateException("Limit <= 0: " + Double.toString(limit));
+        }
+        Optional<Identifiable<?>> identifiableFrom = getIdentifiable(networkRef.get(), from);
+        if (identifiableFrom.isEmpty()) {
+            throw new IllegalStateException("Identifiable from not found: " + from.getId());
+        }
+        Optional<Identifiable<?>> identifiableTo = getIdentifiable(networkRef.get(), to);
+        if (identifiableTo.isEmpty()) {
+            throw new IllegalStateException("Identifiable to not found: " + to.getId());
+        }
+        Terminal terminalFrom = getTerminal(identifiableFrom.get(), from.getSide());
+        Terminal terminalTo = getTerminal(identifiableTo.get(), to.getSide());
+        VoltageAngleLimit voltageAngleLimit = new VoltageAngleLimitImpl(terminalFrom, terminalTo, limit, flowDirection);
+        networkRef.get().getVoltageAngleLimits().add(voltageAngleLimit);
+        return voltageAngleLimit;
+    }
+
+    private static Optional<Identifiable<?>> getIdentifiable(Network network, TerminalRef terminalRef) {
+        return Optional.ofNullable(network.getIdentifiable(terminalRef.getId()));
+    }
+
+    private static Terminal getTerminal(Identifiable<?> identifiable, Side side) {
```

Comment:
For the three last methods, we have to check if we don't have similar code already inside main branch and try to mutualise. TerminalRef is used for regulating terminal already during the serialisation. If not, my opinion is that these methods should not be there, in the voltage angle limit implementation.

Reply:
We moved these methods to the TerminalRef class.